### DISCUSSION TITLE
To-Do List for v1-beta

### DISCUSSION THREAD

**Discussion by sshmatrix** [namesys-eth/ccip2-eth-resources#14] [General]
Here are the things that are remaining to be implemented in the client:

- **IPNS CID generation:** I am having trouble converting the derived ed25519key to IPNS public CID. My output begins with `bafz` in base36 but when converted to base32 with new prefixes, it begins with `k2` instead of `k5`. My exported key in the `w3name` format (multihash 68-byte key) also fails to load, meaning that it is plain wrong and failing checks. I have recreated the entire process of `w3name` package but something is missing. It's just like `isTest` gateways. Something very tiny is off. I will look at it again tomorrow after a break from it.

- **CORS**: Having minor CORS issues with backend. Have all the CORS settings right but still getting opaque response from server over HTTP. Will debug it today/tomorrow.

- **Contract**: Test the contract and deploy it and hook it up the client. @0xc0de4c0ffee is it ready to start testing?

I'll post more updates here

**Comment by 0xc0de4c0ffee** [comment]
> My output begins with bafz in base36 but when converted to base32 with new prefixes, it begins with k2 instead of k5. 

bit confused, base36 is "k" and base32 is "b" prefixed that should be added/removed for base conversion/contenthash.. use js-ipfs/Helia or multibase js?? can we do standalone js code for debugging? it's bit hard to debug with all extra frontend..  

contract should be ok for testnet, it's missing final tests and dev functions.

**Reply by sshmatrix** [reply]
I misspoke. Starts with `bafy` (pure IPNS), `bafz` (IPLD) in base32 and should be `k5` in base36. 

I will make a test script in standalone. I was thinking the same.

**Comment by 0xc0de4c0ffee** [comment]
testing w3name/
IPNS: k51qzi5uqu5dg9jhh7vnq4f56r7qw4jkm51yq71fb12u6vwvgvjb2xip201asr 
Key : 0x08011240b861afd5440711f224f59f45a04ad73b01837f9e1650a770efa688b308321853034a385d144d3c5fe029b5df4c802885d742331bfcd3040779ea3296419844eb

08011240<< not sure what's this prefix, 
it's not matching [github.com/multiformats/multicodec/blob/master/table.csv](https://github.com/multiformats/multicodec/blob/master/table.csv)

08 ? no idea, not in csv
0x01 = v1, 
0x12 = sha256, 
0x40 = length (bytes 32 * 2 = 0x40)

[github.com/libp2p/js-libp2p-crypto/blob/master/src/keys/ed25519.ts#L10](https://github.com/libp2p/js-libp2p-crypto/blob/master/src/keys/ed25519.ts#L10) 
const PRIVATE_KEY_BYTE_LENGTH = 64 // private key is actually 32 bytes but for historical reasons we concat private and public keys

> 0x
prefix : 08011240
A : b861afd5440711f224f59f45a04ad73b01837f9e1650a770efa688b308321853
B: 034a385d144d3c5fe029b5df4c802885d742331bfcd3040779ea3296419844eb

is it privkey+ pubkey or pubkey+privkey? /:P have to test that/

**Reply by sshmatrix** [reply]
I will make the test script and you'll see in there. It's basically this:

<img width="692" alt="Screenshot 2023-04-22 at 19 09 59" src="https://user-images.githubusercontent.com/19473027/233788366-a80155d1-93c0-4f2d-b937-7fcec8d9db53.png">

**Reply by 0xc0de4c0ffee** [reply]
what's `0x70` in that context?

-- | -- | -- | -- | --
-- | -- | -- | -- | --
44 | dag-pb | ipld | 0x70 | permanent | MerkleDAG protobuf
45 | dag-cbor | ipld | 0x71 | permanent | MerkleDAG cbor
46 | libp2p-key | ipld | 0x72 | permanent | Libp2p Public Key

For w3name's import/export key is "0x08011240"+bytes32(private keys) + bytes32(public key) formatted. 
0xe501+0172<prefix+ ipns_lib-p2p key>

**Reply by 0xc0de4c0ffee** [reply]
opened this issue 

https://github.com/paulmillr/noble-curves/issues/35

**Reply by sshmatrix** [reply]
> For w3name's import/export key is "0x08011240"+bytes32(private keys) + bytes32(public key) formatted. 0xe501+0172<prefix+ ipns_lib-p2p key>

Where did you get this formatting? I took the w3name key encoding from here: https://github.com/web3-storage/w3name/blob/main/packages/client/src/index.ts 

^Check lines 75-150. 

In our frontend, I have recreated the same method but it doesn't yield the same result

**Reply by 0xc0de4c0ffee** [reply]
```js
const name1 = await Name.create()
console.log(let key = name1.key.bytes)
const name2 = await Name.from(key)
console.log('Name1:', name1.toString(), 'Name2:', name2.toString())
```

**Reply by 0xc0de4c0ffee** [reply]
`let privateKey = hashToPrivateScalar(hashKey, ed25519.CURVE.n, true).toString(16).padstart(64, "0")`

ed25519 use LE, so `true`.
`hashToPrivateScalar(hash: string | Uint8Array, groupOrder: bigint, isLE?: boolean | undefined): bigint`

ed25519.CURVE.n = 7237005577332262213973186563042994240857116359379907606001950938285454250989 
= 0x1000000000000000000000000000000014DEF9DEA2F79CD65812631A5CF5D3ED, so 62-63 char is ok I guess.. double check, ed25519 keys with full 64 hex chars shouldn't be valid??

comparing secp256k1.curve.n = `0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141` it's too big so we were always getting 64 char.

**Comment by 0xc0de4c0ffee** [comment]
@sshmatrix tested!!

```
        let caip10 = `eip155:1:${App.ADDR}`
        let domain = "domain.eth"
        
        let msg = `Generate IPNS Keys for ${domain}\n\nWARNING:........\n\nSigned By: ${caip10}`;
        let sig = await App.SIGNER.signMessage(msg);
        let password = "pass12#$"
        let inputKey = sha256(
            hexToBytes(
                sig.toLowerCase().startsWith('0x') ? sig.slice(2) : sig
            ) 
        )
        let info = `${caip10}:${domain}`
        let salt = await sha256(`${info}:${password ? password : ''}:${sig.slice(-64)}`)
        let hashKey = await hkdf(sha256, inputKey, salt, info, 42)
        let privateKey = hashToPrivateScalar(hashKey, ed25519.CURVE.n, true).toString(16).padStart(64, "0")
        App.LOG.innerHTML += `<br>privkey: ${privateKey} -- ${privateKey.length}`
        let publicKey = bytesToHex(await ed25519.getPublicKey(privateKey))
        App.LOG.innerHTML += `<br>pubkey${publicKey} -- ${publicKey.length}`
        let key = `08011240${privateKey}${publicKey}`
        const w3Name = await Name.from(hexToBytes(key))
        App.LOG.innerHTML += `<br>IPNS: ${w3Name.toString()}<br>Key: ` + bytesToHex(w3Name.key.bytes)
```

> 
privkey: 0681d6420abb1ba47acd5c03c8e5ee84185a2673576b262e234e50c46d86f597 -- 64
pubkey : 12c8299ec2c51dffbbcb4f9fccadcee1424cb237e9b30d3cd72d47c18103689d -- 64
IPNS: k51qzi5uqu5dgnfwbc46une4upw1vc9hxznymyeykmg6rev1513yrnbyrwmmql
Key: 080112400681d6420abb1ba47acd5c03c8e5ee84185a2673576b262e234e50c46d86f59712c8299ec2c51dffbbcb4f9fccadcee1424cb237e9b30d3cd72d47c18103689d

**Reply by sshmatrix** [reply]
It was that easy?! I was doing extreme shit. Dear lord 🙄 
Thank you so much 🖖

**Comment by sshmatrix** [comment]
> * **IPNS CID generation:** I am having trouble converting the derived ed25519key to IPNS public CID. My output begins with `bafz` in base36 but when converted to base32 with new prefixes, it begins with `k2` instead of `k5`. My exported key in the `w3name` format (multihash 68-byte key) also fails to load, meaning that it is plain wrong and failing checks. I have recreated the entire process of `w3name` package but something is missing. It's just like `isTest` gateways. Something very tiny is off. I will look at it again tomorrow after a break from it.

Fixed!

> * **CORS**: Having minor CORS issues with backend. Have all the CORS settings right but still getting opaque response from server over HTTP. Will debug it today/tomorrow.

Fixed!

> * **Contract**: Test the contract and deploy it and hook it up the client. @0xc0de4c0ffee is it ready to start testing?

In progress
